[CostModel] Mark ssa_copy as free

davemgreen opened the pull request:
These are intrinsics are only used ephemerally and be should be given a zero cost.

fhahn commented:
I am curious if you encountered cases where those intrinsics are still around after the middle-end optimization pipeline. I think all users of those intrinsics should clean them up, but there might be some places that got missed.

davemgreen commented:
No the costing is done during a pass that adds and removes them. This could be handled specially in that pass, but it is more general fix to give a zero cost to everywhere.

nikic commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -1510,6 +1510,8 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
     if (Function::isTargetIntrinsic(IID))
       return TargetTransformInfo::TCC_Basic;
+    if (IID == Intrinsic::ssa_copy)
+      return TargetTransformInfo::TCC_Free;
```
This should be in TargetTransformInfoImpl.h.

davemgreen replied:
Ah I see. That is better, thanks.

davemgreen committed:
[CostModel] Mark ssa_copy as free

These are intrinsics are only used ephemerally and be should be given a zero cost.

nikic reviewed (approved):
LGTM
